### PR TITLE
Make DockerSupportService configuration cache compatible

### PR DESCRIPTION
**Pull request by breskeby**

### CODE REVIEW

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/docker/DockerCommandResult.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+package org.elasticsearch.gradle.internal.docker;
+
+import java.util.Objects;
+
+/**
+ * This class models the result of running a command. It captures the exit code, standard output and standard error and allows
+ * applying String filter for stdout as this is intended to create configuration cache compatible output which
+ * aims to be agnostic.
+ */
+public class DockerCommandResult {
```
Comment (breskeby) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/docker/DockerCommandResult.java:18]: java records are unfortunately not supported by gradle here yet

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/docker/DockerValueSource.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            // The redundant cast is to silence a compiler warning.
+            spec.setCommandLine(args);
+            spec.setStandardOutput(stdout);
+            spec.setErrorOutput(stderr);
+            spec.setIgnoreExitValue(true);
+        });
+        return new DockerResult(execResult.getExitValue(), filtered(stdout.toString()), stderr.toString());
+    }
+
+    private String filtered(String input) {
```
Comment (breskeby) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/docker/DockerValueSource.java:68]: that allows us to "normalise" the output to make configuration cache reuse more likely

**Review by mark-vieira — APPROVED** [review]
LGTM.

How can we be aware of everywhere we need to implement a custom `ValueSource`? Certainly there are other places we do this, like in `ParallelDetector` and `GitInfo`.

**Comment by breskeby** [comment]
@mark-vieira there is no really good tooling around this. So far this is a manual process and often trial and error running builds with --configuration-cache enabled. In this case here I started with just using a cc compatible exec command but then seeing that the configuration cache was never up-to-date due to changing output in different docker commands. that lead to ValueSource which allows additional tweaking like filtering the command output

**Comment by mark-vieira** [comment]
Ah I see. So we only really need to implement a `ValueSource` where we want better control around input/output snapshotting to get better cache reuse. Makes sense.

**Comment by breskeby** [comment]
This work was related to https://github.com/elastic/elasticsearch/issues/57918

**Pull request merged**